activerecord/lib/active_record/collection_cache_key.rb

```diff
@@ -2,30 +2,15 @@ module ActiveRecord
   module CollectionCacheKey
 
     def collection_cache_key(collection = all, timestamp_column = :updated_at) # :nodoc:
-      query_signature = Digest::MD5.hexdigest(collection.to_sql)
-      key = "#{collection.model_name.cache_key}/query-#{query_signature}"
+      model_signature = collection.model_name.cache_key
 
       if collection.loaded?
-        size = collection.size
-        timestamp = collection.max_by(&timestamp_column).public_send(timestamp_column)
+        unique_signature = collection.pluck(primary_key, timestamp_column).flatten.join("-".freeze)
       else
-        column_type = type_for_attribute(timestamp_column.to_s)
-        column = "#{connection.quote_table_name(collection.table_name)}.#{connection.quote_column_name(timestamp_column)}"
-
-        query = collection
-          .select("COUNT(*) AS size", "MAX(#{column}) AS timestamp")
-          .unscope(:order)
-        result = connection.select_one(query)
-
-        size = result["size"]
-        timestamp = column_type.deserialize(result["timestamp"])
+        unique_signature = collection.unscope(:order).pluck(primary_key, timestamp_column).flatten.join("-".freeze)
```

Comment:
What do you mean by PostgreSQL order failures? If the scope has an order then the cache key query has to preserve it otherwise it won't match the same set of records.

```
[3] pry(main)> Post
=> Post(id: integer, published_at: datetime, created_at: datetime, updated_at: datetime)
[4] pry(main)> 20.times { |i| Post.create!(published_at: Time.current - 1.hour * i) }
[5] pry(main)> Post.count
   (0.1ms)  SELECT COUNT(*) FROM "posts"
=> 20
[6] pry(main)> Post.order(published_at: :desc).limit(10).pluck(:id, :updated_at)
   (0.3ms)  SELECT  "posts"."id", "posts"."updated_at" FROM "posts" ORDER BY "posts"."published_at" DESC LIMIT 10
=> [[1, Wed, 25 Nov 2015 21:40:23 UTC +00:00],
 [2, Wed, 25 Nov 2015 21:40:23 UTC +00:00],
 [3, Wed, 25 Nov 2015 21:40:23 UTC +00:00],
 [4, Wed, 25 Nov 2015 21:40:23 UTC +00:00],
 [5, Wed, 25 Nov 2015 21:40:23 UTC +00:00],
 [6, Wed, 25 Nov 2015 21:40:23 UTC +00:00],
 [7, Wed, 25 Nov 2015 21:40:23 UTC +00:00],
 [8, Wed, 25 Nov 2015 21:40:23 UTC +00:00],
 [9, Wed, 25 Nov 2015 21:40:23 UTC +00:00],
 [10, Wed, 25 Nov 2015 21:40:23 UTC +00:00]]
[7] pry(main)> Post.order(published_at: :asc).limit(10).pluck(:id, :updated_at)
   (0.2ms)  SELECT  "posts"."id", "posts"."updated_at" FROM "posts" ORDER BY "posts"."published_at" ASC LIMIT 10
=> [[20, Wed, 25 Nov 2015 21:40:23 UTC +00:00],
 [19, Wed, 25 Nov 2015 21:40:23 UTC +00:00],
 [18, Wed, 25 Nov 2015 21:40:23 UTC +00:00],
 [17, Wed, 25 Nov 2015 21:40:23 UTC +00:00],
 [16, Wed, 25 Nov 2015 21:40:23 UTC +00:00],
 [15, Wed, 25 Nov 2015 21:40:23 UTC +00:00],
 [14, Wed, 25 Nov 2015 21:40:23 UTC +00:00],
 [13, Wed, 25 Nov 2015 21:40:23 UTC +00:00],
 [12, Wed, 25 Nov 2015 21:40:23 UTC +00:00],
 [11, Wed, 25 Nov 2015 21:40:23 UTC +00:00]]
[8] pry(main)> Post.order(published_at: :desc).limit(10).cache_key
   (0.2ms)  SELECT  "posts"."id", "posts"."updated_at" FROM "posts" LIMIT 10
=> "posts/collection-digest-7f2daae36b230c64b72b28cafd70ddbd07d0f6369fb46e3155cc90ebcdad5058"
[9] pry(main)> Post.order(published_at: :asc).limit(10).cache_key
   (0.1ms)  SELECT  "posts"."id", "posts"."updated_at" FROM "posts" LIMIT 10
=> "posts/collection-digest-7f2daae36b230c64b72b28cafd70ddbd07d0f6369fb46e3155cc90ebcdad5058"
```